app/addons/documents/doc-editor/components.react.jsx

```diff
@@ -0,0 +1,540 @@
+        keyChecked.push('_rev');
+      }
+      if (_.isEmpty(_.difference(keyChecked, _.keys(json)))) {
+        return;
+      }
+
+      this.getEditor().setReadOnly(true);
+      setTimeout(function () { this.getEditor().setReadOnly(false); }.bind(this), 400);
+
+      // extend ensures _id stays at the top of the editor doc
+      json = _.extend({ _id: this.state.doc.id, _rev: this.state.doc.get('_rev') }, json);
+      this.getEditor().setValue(JSON.stringify(json, null, '  '));
+
+      FauxtonAPI.addNotification({
+        type: 'error',
+        msg: "Cannot remove a document's id or revision.",
+        clear: true
+      });
+    },
+
+    componentDidMount: function () {
+      store.on('change', this.onChange, this);
+    },
+
+    componentWillUnmount: function () {
+      store.off('change', this.onChange);
+    },
+
+    onChange: function () {
+      if (this.isMounted()) {
+        this.setState(this.getStoreState());
+      }
+    },
+
+    saveDoc: function () {
+      Actions.saveDoc(this.state.doc, this.checkDocIsValid(), this.onSaveComplete);
+    },
+
+    onSaveComplete: function (json) {
+      this.getEditor().setValue(json);
+      this.getEditor().clearChanges();
+
+      // the save action updates the doc. This ensures the button row then shows the appropriate buttons
+      this.forceUpdate();
+    },
+
+    hideDeleteDocModal: function () {
+      Actions.hideDeleteDocModal();
+    },
+
+    deleteDoc: function () {
+      Actions.hideDeleteDocModal();
+      Actions.deleteDoc(this.state.doc);
+    },
+
+    getEditor: function () {
+      return (this.refs.docEditor) ? this.refs.docEditor.getEditor() : null;
+    },
+
+    checkDocIsValid: function () {
+      if (this.getEditor().hasErrors()) {
+        return false;
+      }
+      var json = JSON.parse(this.getEditor().getValue());
+      this.state.doc.clear().set(json, { validate: true });
+
+      return !this.state.doc.validationError;
+    },
+
+    clearChanges: function () {
+      this.refs.docEditor.clearChanges();
+    },
+
+    getButtonRow: function () {
+      if (this.props.isNewDoc) {
+        return false;
+      }
+      return (
+        <div>
+          <AttachmentsPanelButton doc={this.state.doc} isLoading={this.state.isLoading} />
+          <div className="doc-editor-extension-icons"></div>
+          <PanelButton title="Upload Attachment" iconClass="icon-circle-arrow-up" onClick={Actions.showUploadModal} />
+          <PanelButton title="Clone Document" iconClass="icon-repeat" onClick={Actions.showCloneDocModal} />
+          <PanelButton title="Delete" iconClass="icon-trash" onClick={Actions.showDeleteDocModal} />
+        </div>
+      );
+    },
+
+    uploadComplete: function () {
+      Actions.initDocEditor({
+        doc: this.state.doc,
+        onLoaded: function () {
+          this.getEditor().setValue(JSON.stringify(this.state.doc.attributes, null, '  '));
+          this.refs.uploadModal.reset();
```

Comment:
Very good, I didn't think of that. I think the parent component (which passes the onSuccess method to the upload component) could listen to a published "done" msg and do whatever it needs there. Much cleaner.